indexing-service/src/main/java/org/apache/druid/indexing/common/task/RangePartitionCachingLocalSegmentAllocator.java

```diff
@@ -0,0 +1,195 @@
+
+/**
+ * Allocates all necessary range-partitioned segments locally at the beginning and reuses them.
+ *
+ * @see CachingLocalSegmentAllocator
+ */
+public class RangePartitionCachingLocalSegmentAllocator implements IndexTaskSegmentAllocator
+{
+  private final String dataSource;
+  private final String partitionDimension;
+  private final Map<Interval, String[]> intervalsToPartitions;
+  private final IndexTaskSegmentAllocator delegate;
+
+  public RangePartitionCachingLocalSegmentAllocator(
+      TaskToolbox toolbox,
+      String taskId,
+      String supervisorTaskId,
+      String dataSource,
+      String partitionDimension,
+      Map<Interval, String[]> intervalsToPartitions
+  ) throws IOException
+  {
+    this.dataSource = dataSource;
+    this.partitionDimension = partitionDimension;
+    this.intervalsToPartitions = intervalsToPartitions;
+
+    this.delegate = new CachingLocalSegmentAllocator(
+        toolbox,
+        taskId,
+        supervisorTaskId,
+        this::getIntervalToSegmentIds
+    );
+  }
+
+  private Map<Interval, List<SegmentIdWithShardSpec>> getIntervalToSegmentIds(Function<Interval, String> versionFinder)
+  {
+    Map<Interval, List<SegmentIdWithShardSpec>> intervalToSegmentIds =
+        Maps.newHashMapWithExpectedSize(intervalsToPartitions.size());
+
+    intervalsToPartitions.forEach(
+        (interval, partitions) ->
+            intervalToSegmentIds.put(
+                interval,
+                translatePartitions(interval, partitions, versionFinder)
+            )
+    );
+
+    return intervalToSegmentIds;
+  }
+
+  private List<SegmentIdWithShardSpec> translatePartitions(
+      Interval interval,
+      String[] partitions,
+      Function<Interval, String> versionFinder
+  )
+  {
+    if (partitions.length == 0) {
+      return Collections.emptyList();
+    }
+
+    String[] uniquePartitions = Arrays.stream(partitions).distinct().toArray(String[]::new);
+    int numUniquePartition = uniquePartitions.length;
+
+    if (numUniquePartition == 1) {
+      return Collections.singletonList(
+          createLastSegmentIdWithShardSpec(
+              interval,
+              versionFinder.apply(interval),
+              uniquePartitions[0],
+              0
+          )
+      );
+    }
+
+    if (isLastPartitionOnlyMaxValue(partitions)) {
+      // The last partition only contains the max value. A shard that just contains the max value is likely to be
+      // small, so combine it with the second to last one.
+      numUniquePartition -= 1;
+    }
+
+    List<SegmentIdWithShardSpec> segmentIds =
+        IntStream.range(0, numUniquePartition - 1)
+                 .mapToObj(i -> createSegmentIdWithShardSpec(
+                     interval,
+                     versionFinder.apply(interval),
+                     uniquePartitions[i],
+                     uniquePartitions[i + 1],
+                     i
+                 ))
+                 .collect(Collectors.toCollection(ArrayList::new));
```

Comment:
The [javadoc for Collectors.toList()](https://docs.oracle.com/javase/8/docs/api/java/util/stream/Collectors.html#toList--) states:

> There are no guarantees on the type, mutability, serializability, or thread-safety of the List returned

Since the returned `List` needs to be mutated, I specified the list implementation used by the `Collector`.